Add async support

kevinheavey opened the pull request:
Closes #80 

Summary of changes:

- Adds `AsyncClient` and `AsyncToken` classes.
- Refactors sync code so the async code can use it with minimal duplication. No changes to logic or behaviour of sync methods.
- Adds tests for async methods.
- Adds async example to the README
- Adds httpx as a dependency.
- Fixes `test_get_identity` which I noticed wasn't testing `get_identity`. Instead it was redundantly testing `get_genesis_hash`.
- Bumps minor version number.

kevinheavey commented:
Should be good to run the CI again. By the way, any idea why those two docstrings didn't show up when I ran `make lint`? I'm pretty sure that command caught similar errors for me before.

michaelhly commented on solana/rpc/api.py:
```diff
@@ -67,13 +401,1039 @@ def get_balance(self, pubkey: Union[PublicKey, str], commitment: Optional[Commit
+        return resp
+
+
+class AsyncClient(_ClientCore):  # pylint: disable=too-many-public-methods
```
Would it be possible to split into three different files?

- solana/rpc/core.py
- solana/rpc/api.cpy (old synchornous client)
- solana/rpc/async_api.py (new async client)

kevinheavey replied:
Will do

kevinheavey replied:
In the same spirit, I'll give AsyncToken its own file too

michaelhly commented on solana/rpc/providers/http.py:
```diff
@@ -2,60 +2,123 @@
+
+
+class AsyncHTTPProvider(AsyncBaseProvider, _HTTPProviderCore):
+    """Async HTTP provider to interact with the http rpc endpoint."""
```
Same here?

michaelhly commented on tests/integration/test_http_client.py:
```diff
@@ -2,18 +2,25 @@
 
 @pytest.mark.integration
-def test_request_air_drop(stubbed_sender, test_http_client):
+def test_request_air_drop(stubbed_sender, test_http_clients):
```
Can we test the async client separate from the old client? Having both synchronous and async tests in the same method makes maintenance more difficult.

kevinheavey replied:
This is possible, but we'll have a lot of duplicated code without a big refactor of the test suite. Plus there are places where we just test that the response the async client received is the same as that of the sync client, and we'll lose that (without more refactoring).

To get this over the line I'll just duplicate all the old pre-async tests and make async versions of them. Your call if that's what you want

michaelhly replied:
Yes. Duplicate tests are fine, it's just less confusing. We can decide to refactor and try to reduce duplicates later.

michaelhly reviewed (commented):
Hey @kevinheavey. Thank you for contributing. This PR looks great. 

I think it would be better if we have clearer boundaries between the old synchronous client and the new async client so that this codebase is easier to maintain. 

And that if someone decides implement a websocket client, they should not be cramming into the existing `api.py` file nor the existing integration tests.

kevinheavey commented:
Made those changes.
Extras:
- Removed pytest-tornasync as a dev dependency. It wasn't doing anything until I added async tests using pytest-asyncio, at which point it started confusing pytest.
- Fixed typo in tests: reciever -> receiver

kevinheavey commented:
Forgot the linters, should work now

kevinheavey commented:
Think pylint should be okay now

michaelhly commented:
Let's just disable `R0913: Too many arguments (7/5) (too-many-arguments)` for those modules. 

https://pylint.readthedocs.io/en/latest/faq.html#is-there-a-way-to-disable-a-message-for-a-particular-module-only

kevinheavey commented:
Done. Got a good feeling about this one. Pylint is not complaining about anything I touched

michaelhly reviewed (approved):
Thanks, @kevinheavey. This is great!
